espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/dispatcher/PointerDispatch.java

```diff
@@ -219,69 +219,82 @@ public static void dispatchPointerCancel(final W3CActionAdapter dispatcherAdapte
-                            pointerInputState.setX(x);
-                            pointerInputState.setY(y);
-
-                            if (!isLast) {
-                                // 10. Asynchronously wait for an implementation defined amount of time to pass
-                                dispatcherAdapter.sleep(dispatcherAdapter.pointerMoveIntervalDuration());
-                            }
-                        } finally {
-                            dispatcherAdapter.unlockAdapter();
-                        }
-                    }
 
-                    if (!isLast) {
+                // 2. Let time delta be the time since the beginning of the current tick, measured in milliseconds on a monotonic clock
+                long timeDelta = System.currentTimeMillis() - timeSinceBeginningOfTick;
+
+                // 3. Let duration ratio be the ratio of time delta and duration, if duration is greater than 0, or 1 otherwise
+                float durationRatio = duration > 0 ? timeDelta / ((float) duration) : 1;
+
+                // 4. If duration ratio is 1, or close enough to 1 that the implementation will not further subdivide the move action,
+                //    let last be true. Otherwise let last be false
+                isLast = (1 - durationRatio) <= dispatcherAdapter.getPointerMoveDurationMargin(pointerInputState);
+
+                // 5. If last is true, let x equal target x and y equal target y
+                // 6. Otherwise let x equal an approximation to duration ratio × (target x - start x) + start x,, ...
+                final long x = isLast ? targetX : Math.round(durationRatio * (targetX - startX)) + startX;
+                final long y = isLast ? targetY : Math.round(durationRatio * (targetY - startY)) + startY;
+
+                // 7-8: Let currentX and currentY be pointer input state
+                final long currentX = pointerInputState.getX();
+                final long currentY = pointerInputState.getY();
+
+                // Prepare the result
+                DispatchPointerMoveResult dispatchResult = new DispatchPointerMoveResult(
+                        dispatcherAdapter,
+                        sourceId, pointerInputState.getType(),
+                        currentX, currentY, x, y,
+                        pointerInputState.getButtons(),
+                        globalKeyInputState,
+                        duration
+                );
+
+                if (currentX != x || currentY != y) {
+
+                    // 8.3. Let input state's x property equal x and y property equal y
+                    pointerInputState.setX(x);
+                    pointerInputState.setY(y);
+                }
+
+                // Sleep for a fixed period of time
+                if (duration > 0) {
+                    try {
                         dispatcherAdapter.sleep(dispatcherAdapter.pointerMoveIntervalDuration());
```

Comment:
it won't throw InterruptedException if we use SystemClock

Reply:
I just changed it to throw an AppiumException and the Dummy Adapter used for testing catches the InterruptedException internally